vllm/entrypoints/openai/api_server.py

```diff
@@ -308,28 +308,35 @@ async def run_server(args, llm_engine=None, **uvicorn_kwargs) -> None:
     logger.info("vLLM API server version %s", VLLM_VERSION)
     logger.info("args: %s", args)
 
-    server = await build_server(
-        args,
-        llm_engine,
-        **uvicorn_kwargs,
-    )
-
-    loop = asyncio.get_running_loop()
-
-    server_task = loop.create_task(server.serve())
-
-    def signal_handler() -> None:
-        # prevents the uvicorn signal handler to exit early
-        server_task.cancel()
-
-    loop.add_signal_handler(signal.SIGINT, signal_handler)
-    loop.add_signal_handler(signal.SIGTERM, signal_handler)
-
     try:
-        await server_task
-    except asyncio.CancelledError:
-        print("Gracefully stopping http server")
-        await server.shutdown()
+        server = await build_server(
+            args,
+            llm_engine,
+            **uvicorn_kwargs,
+        )
+
+        loop = asyncio.get_running_loop()
+
+        server_task = loop.create_task(server.serve())
+
+        def signal_handler() -> None:
+            # prevents the uvicorn signal handler to exit early
+            server_task.cancel()
+
+        loop.add_signal_handler(signal.SIGINT, signal_handler)
+        loop.add_signal_handler(signal.SIGTERM, signal_handler)
+
+        try:
+            await server_task
+        except asyncio.CancelledError:
+            print("Gracefully stopping http server")
+            await server.shutdown()
+    finally:
+        # Clean up globals
+        for var in ("openai_serving_chat", "openai_serving_completion",
+                    "openai_serving_embedding", "openai_serving_tokenization",
+                    "engine_args", "engine"):
+            globals().pop(var, None)
```

Comment:
does `del` work? `globals().pop` is too hacky.

Reply:
@youkaichao we'd have to check and `del` each one individually so it would be a lot more code. I want this to work whether or not each is defined, in case some error occurs after setting some and not others.

The way globals are used here is already hacky imo and I think we'll clean it up later. I wanted to keep this change as simple as possible as there are overlapping changes in https://github.com/vllm-project/vllm/pull/6883 which will be merged very soon.

Reply:
got it. this is a minor concern, and you can skip it if it is difficult to solve.

the most important part is still make the ci pass 🙏